constantine/proof_systems/constraint_systems/r1cs_circom_parser.nim

```diff
@@ -83,23 +149,117 @@ proc parseConstraint(f: File, constraint: var Constraint, fieldSize: int32): boo
   ?f.parseLinComb(constraint.C, fieldSize)
   return true
 
-template r1csSection(body: untyped): untyped =
-  var sectionSize: int64
-  ?f.parseInt(sectionSize, littleEndian)
+template r1csSection(sectionSize, body: untyped): untyped =
   let startOffset = f.getFilePosition()
 
   body
 
-  return sectionSize == f.getFilePosition() - startOffset
+  return sectionSize.int == f.getFilePosition() - startOffset
 
-proc parseConstraints(f: File, constraints: var seq[Constraint], numConstraints, fieldSize: int32): bool =
-  r1csSection:
+proc parseConstraints(f: File, constraints: var seq[Constraint], sectionSize: uint64, numConstraints, fieldSize: int32): bool =
+  r1csSection(sectionSize):
     constraints.setLen(numConstraints)
     for constraint in constraints.mitems():
       ?f.parseConstraint(constraint, fieldSize)
 
-proc parseWireLabelMap(f: File, wireIdToLabelId: var seq[int64], numWires: int32): bool =
-  r1csSection:
-    wireIdToLabelId.setLen(numWires)
-    for labelId in wireIdToLabelId.mitems():
+proc parseMagicHeader(f: File, mh: var array[4, char]): bool =
+  let num = f.readChars(mh, 0, 4) # from start of file!
+  result = num == 4
+
+proc parseSectionKind(f: File, v: var R1csSectionKind): bool =
+  var val: uint32
+  result = f.parseInt(val, littleEndian)
+  # convert from uint32 value and correct for -1 in definition of enum
+  v = R1csSectionKind(val.int - 1)
+
+proc parseHeader(f: File, h: var Header): bool =
+  ?f.parseInt(h.fieldSize, littleEndian) # byte size of the prime number
+  # allocate for the prime
+  h.prime = newSeq[byte](h.fieldSize)
+  ?f.readInto(h.prime)
+  ?f.parseInt(h.nWires, littleEndian)
+  ?f.parseInt(h.nPubOut, littleEndian)
+  ?f.parseInt(h.nPubIn, littleEndian)
+  ?f.parseInt(h.nPrvIn, littleEndian)
+  ?f.parseInt(h.nLabels, littleEndian)
+  ?f.parseInt(h.nConstraints, littleEndian)
+  result = true # would have returned before due to `?` otherwise
+
+proc parseWire2Label(f: File, v: var Wire2Label, sectionSize: uint64): bool =
+  r1csSection(sectionSize):
+    let numWires = sectionSize div 8
+    v.wireIds.setLen(numWires)
+    for labelId in v.wireIds.mitems():
       ?f.parseInt(labelId, littleEndian)
+
+## TODO: to be written :)
```

Comment:
Note that I am unaware of any usage of these so they are not in scope for this PR